### PR TITLE
Support for static syntax without object instance methods

### PR DESCRIPTION
**Pull request by saivann**
This PR allows static functions to continue working when overriding reserved keys such as `setIn` with custom data in the objects.

Previously custom data would be dropped. Now custom data would override instance functions. Neither are good, which is why this PR is including all steps to migrate existing users to the new syntax.

Backward compatibility is maintained, but users using the old syntax would get this kind of warning
`DEPRECATED: Using x.asObject(...) will be deprecated. Please use Immutable.asObject(x, ...) instead.`

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -72,7 +72,8 @@ Immutable([1, 2, 3]).concat([10, 9, 8]).sort()
 // [1, 2, 3, 4, 5, 6], because a vanilla array's concat() method has
 // no knowledge of Immutable.
 
-Immutable({all: "your base", are: {belong: "to them"}}).merge({are: {belong: "to us"}})
+var obj = Immutable({all: "your base", are: {belong: "to them"}});
+Immutable.merge(obj, {are: {belong: "to us"}})
 // This handy new method will return the following:
```
Comment (suchipi) [inline README.md]: It's not really a "method" per se anymore, so maybe the wording should be changed here

File: README.md [hunk 2]
```diff
@@ -318,14 +324,15 @@ Multiple keys can be provided, either in an Array or as extra arguments.
...
+var obj = Immutable({when: "the", levee: "breaks"});
+var mutable = Immutable.asMutable(obj);
 
 mutableObject.have = "no place to go";
 
 mutableObject // {when: "the", levee: "breaks", have: "no place to go"}
 ```
 
-Returns a mutable copy of the object. For a deeply mutable copy, in which any instances of `Immutable` contained in nested data structures within the object have been converted back to mutable data structures, call `.asMutable({deep: true})` instead.
+Returns a mutable copy of the object. For a deeply mutable copy, in which any instances of `Immutable` contained in nested data structures within the object have been converted back to mutable data structures, call `.asMutable(obj, {deep: true})` instead.
```
Comment (suchipi) [inline README.md]: I think this was written as `.asMutable` because the object to the left of the dot could be anything; with the docs only talking about the static method syntax, this could be changed to `Immutable.asMutable`

File: src/seamless-immutable.js [hunk 3]
```diff
@@ -5,11 +5,36 @@
...
+      if (!prototype) {
+          return {};
+      } else {
+          return Object.create(prototype);
+      }
+  }
+
+  function staticWarning(fn, name) {
+      function staticWrapper() {
+          console.log('DEPRECATED: Using x.' + name + '(...) will be deprecated. Please use Immutable.' + name + '(x, ...) instead.');
```
Comment (suchipi) [inline src/seamless-immutable.js]: It would be nice to use console.warn instead.

Also, instead of "will be deprecated", maybe it would be better to say "has been deprecated". Saying "DEPRECATED: ...will be deprecated" is a bit confusing.

File: src/seamless-immutable.js [hunk 4]
```diff
@@ -5,11 +5,36 @@
...
+          return fn.apply(this, [].slice.call(arguments));
+      }
+
+      if (process.env.NODE_ENV !== "production") {
```
Comment (suchipi) [inline src/seamless-immutable.js]: Will this blow up in the browser? Not sure what the build system for this project is like

**Comment by saivann** [comment]
@suchipi Thanks, just applied all requested changes and added more test units.

**Comment by saivann** [comment]
@rtfeldman Note; The migration to the new syntax is optional. I can easily drop the console.warn warnings from this PR if you prefer to keep supporting both syntax going forward.

**Comment by rtfeldman** [comment]
After a lot of thought, I don't think outright migrating to this syntax is the right move. I think it would be better to make this opt-in, for example:

```js
var Immutable = require("seamless-immutable").static;
```

Making this give you a new version of `Immutable` that doesn't override methods seems like the best way to go. It preserves backwards compatibility and would be easy to opt into.

What do you think?

**Comment by saivann** [comment]
@rtfeldman I have dropped the deprecation warning from this pull request such that it is now mostly only #130 bugfix and test units for it, and not a migration scenario.

This PR still includes [https://github.com/rtfeldman/seamless-immutable/pull/160/commits/2ca121b942c98638531c8df6fca5d5b85bdfbfad](this breaking change). However in my views, losing data or losing access to the method are both undesirable bugs, and this is the key to allow the user to easily switch to a static syntax that doesn't have the bug if needs be. Thus in my humble opinion merging this would be an improvement.

I gave a good hour at exploring the idea of `require("seamless-immutable").static;` to address #170. This seems like a good idea at first, however this appeared non-trivial to me, as `Immutable()`, `makeImmutableArray()`, `makeImmutableObject()` and such are called generously from within the library. It appears pretty easy to miss a case and introduce bugs, or re-pollute the object.

Although I do understand that it is great to support previous syntax, I am generally a fan of keeping things simple to ease maintainability and avoid bugs. It seems like there is no argument in favor of the previous syntax beside that it's currently in use, the static syntax appears superior on all fronts.

**Comment by saivann** [comment]
tl;dr; I suggest we only merge bugfixes, and choose between migrating or not migrating on a separate PR.

**Comment by saivann** [comment]
@rtfeldman I may have found a simple compromise that is similar to what you are suggesting. I pushed another commit for this

https://github.com/saivann/seamless-immutable/tree/collision#static-or-instance-syntax

```
var Immutable = require("seamless-immutable");
Immutable.config({
    use_instance_methods: false
});
```

**Comment by MoeSattler** [comment]
@saivann would the config be opt-in or opt-out?

**Comment by saivann** [comment]
@MoeSattler, the config only defines if seamless-immutable should add sub-methods to all objects it creates. This is opt-in. In summary, in this PR:

1. Seamless-immutable continues to support the same syntaxes it already supports
```
Immutable.setIn(obj, 'key', data)
```
```
obj.setIn('key', data)
```

2. In case of collision, user data now has priority over seamless-immutable methods:

Previously:
```
var obj = Immutable({setIn: 'data'})
obj.setIn; // function, the assigned data is lost
```

Now:
```
var obj = Immutable({setIn: 'data'})
obj.setIn; // 'data', the seamless-immutable method is lost
```

3. User can opt-in to attach seamless-immutable methods to objects

```
var Immutable = require("seamless-immutable");
Immutable.config({
    use_instance_methods: false
});

var obj = Immutable({})
obj.setIn; // undefined, the object is clean
```

**Comment by suchipi** [comment]
To remain completely backwards compatible ("bugwards compatible"), the second item in your summary would need to be reverted to the old behavior; for users who need a workaround for data loss, I think the static methods should be recommended instead (or a config key could be added to prioritize user data instead of methods).

Even though prioritizing user data over instance methods is a nicer API in an ideal world... maintaining backwards compatibility and avoiding a major version bump are important from a package maintainer's perspective, too.

**Comment by saivann** [comment]
@suchipi Although I can easily make this conditional to the `use_instance_methods` config if I am asked for it, I think this is actually a bad idea that I'd like to discourage, because it makes it harder for users to have bug-free library by default. Users using the static syntax would still be affected unless they also set `use_instance_methods`, which I think we can safely assume most users won't.

And perhaps obvious, but it seems unnecessary to me that we keep backward compatibility with a bug. There is no perfect backward compatible solution for the existing syntax. But we can however make sure the library is bug-free for new users, while also having a solution at hands for any existing users who may be affected this bug. This is what I am recommending in this PR.

**Comment by saivann** [comment]
Actually, I am changing my mind on this ^ . There are security implications with allowing users to run an instance method, while at the same time allowing arbitrary override those methods. I'll update the PR and try to come up with something better.

**Comment by saivann** [comment]
OK latest attempt:

1. Seamless-immutable continues to support the same syntaxes it already supports
```
Immutable.setIn(obj, 'key', data)
```
```
obj.setIn('key', data)
```

2. In case of collision, user data is lost, no matter what syntax (no change with current spec):

```
var obj = Immutable({setIn: 'data'})
obj.setIn; // function, the assigned data is lost
```

3. User can opt-in to stop attaching seamless-immutable methods to objects

```
var Immutable = require("seamless-immutable").init({
    use_static: true
});

var obj = Immutable({})
obj.setIn; // undefined, the object is clean
```

This is pretty in line to what @rtfeldman asked (sorry for all the extra loops before getting there!).

I still think it'd be great to migrate to one syntax in the future, but can be for later.

**Comment by rtfeldman** [comment]
@saivann This is great! 😍 Thank you so much!

My only request would be to simplify it to just:

```js
var Immutable = require("seamless-immutable").static;
```

I don't anticipate adding more configuration options later, so having an init function feels like overkill.

Sound good?

**Comment by saivann** [comment]
@rtfeldman Just added a commit for this, thanks! Looking forward to use seamless-immutable 6.4.0 :)

**Comment by MoeSattler** [comment]
Love this! Thanks so much!

**Comment by rtfeldman** [comment]
@saivann Thanks! I'm going to pair this with #177 and release as 7.0.0

**Comment by rtfeldman** [comment]
Released as 7.0.0.

**Comment by draunkin** [comment]
Thanks so much for this.  I was developing some reporting in our application which uses data that is not frequently accessed and only then did I realise this was happening to some of my objects which have since been serialized into the db.
... well can't reverse time but so glad you've already got a solution for it!
